src/main/java/com/saucelabs/ci/sauceconnect/AbstractSauceTunnelManager.java

```diff
@@ -761,65 +718,101 @@ private int findFreePort() throws SauceConnectException {
+    private final int port;
+    private final Logger LOGGER;
 
     private String tunnelId;
     private boolean failed;
+    private boolean apiResponse;
+
+    private HttpClient client = HttpClient.newHttpClient();
+    private static final int sleepTime = 1000;
 
-    public SystemOutGobbler(
+    public SCMonitor(
         String name,
-        InputStream is,
-        final Semaphore semaphore,
-        PrintStream printStream,
-        String startedMessage) {
-      super(name, is, printStream);
-      this.semaphore = semaphore;
-      this.startedMessage = startedMessage;
+        final int port,
+        final Logger logger) {
+      super(name);
+      this.port = port;
+      this.LOGGER = logger;
     }
 
-    /**
-     * {@inheritDoc}
-     *
-     * <p>If the line contains the Sauce Connect started message, then release the semaphone, which
-     * will allow the build to resume.
-     *
-     * @param line Line being processed
-     */
-    @Override
-    protected void processLine(String line) {
-      super.processLine(line);
-
-      System.out.println(line);
-      if (StringUtils.containsIgnoreCase(line, "sauce connect running id=")) {
-        tunnelId = StringUtils.substringAfter(line, "sauce connect running id=").trim();
-      }
-      if (StringUtils.containsIgnoreCase(line, "fatal error exiting")) {
-        failed = true;
-      }
-      if (StringUtils.containsIgnoreCase(line, startedMessage) || failed) {
-        // unlock processMonitor
-        semaphore.release();
-      }
+    public void setSemaphore(Semaphore semaphore) {
+      this.semaphore = semaphore;
     }
 
     public String getTunnelId() {
-      return tunnelId;
+      HttpRequest request = HttpRequest.newBuilder()
+                        .uri(URI.create(String.format("http://localhost:%d/info", port)))
+                        .GET()
+                        .build();
+      try {
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+        String responseBody = response.body();
+        JSONObject jsonObject = new JSONObject(responseBody);
+        if (jsonObject.has("tunnel_id")) {
+          return jsonObject.getString("tunnel_id");
+        }
+      } catch (Exception e) {
+        this.LOGGER.info("Failed to get tunnel id", e);
+        return null;
+      }
+      this.LOGGER.info("Failed to get tunnel id");
+      return null;
     }
 
     public boolean isFailed() {
       return failed;
     }
-  }
 
-  /** Handles processing Sauce Connect output sent to stderr. */
-  public class SystemErrorGobbler extends StreamGobbler {
+    public void run() {
+      while (true) {
+        pollEndpoint();
+        if (this.semaphore.availablePermits() > 0) {
+          return;
+        }
+
+        try {
+          Thread.sleep(sleepTime);
+        } catch ( java.lang.InterruptedException e ) {
+          return;
+        }
+      }
+    }
+
+    private void pollEndpoint() {
```

Comment:
Here we could be using `/readyz` endpoint it would do all the checks for you.
The `/status` is informational and there is also `/info`.

Reply:
Fixed in 991382c